File: functions/src/boost/boostActivationService.ts
import {ALLOW_STACKING, BOOST_PRODUCTS} from "./config.js";
import type {ActiveBoostSnapshot} from "./types.js";

export type ActivationDecision =
  | {shouldActivate: true; startedAt: Date; expiresAt: Date; extendBoostId?: string}
  | {shouldActivate: false; alreadyActive: true}
  | {shouldActivate: false; insufficientBalance: true};

export class BoostActivationService {
  constructor(
    private readonly allowStacking = ALLOW_STACKING,
    private readonly durationMs = BOOST_PRODUCTS.durationMs,
  ) {}

  activeBoost(boosts: ActiveBoostSnapshot[], now: Date): ActiveBoostSnapshot | null {
    for (const boost of boosts) {
      if (this.isActive(boost, now)) {
        return boost;
      }
    }
    return null;
  }

  isActive(boost: ActiveBoostSnapshot, now: Date): boolean {
    if (boost.status !== "active") {
      return false;
    }
    if (!boost.expiresAt) {
      return false;
    }
    return boost.expiresAt.getTime() > now.getTime();
  }

  decide(params: {
    now: Date;
    currentActive: ActiveBoostSnapshot | null;
    balance?: number;
    durationMs?: number;
    requireBalance?: boolean;
  }): ActivationDecision {
    const grantMs = params.durationMs ?? this.durationMs;
    if (!Number.isFinite(grantMs) || grantMs <= 0) {
      return {shouldActivate: false, insufficientBalance: true};
    }
    const live =
      params.currentActive && this.isActive(params.currentActive, params.now)
        ? params.currentActive
        : null;
    if (live && !this.allowStacking) {
      return {shouldActivate: false, alreadyActive: true};
    }
    if (params.requireBalance && (params.balance ?? 0) < 1) {
      return {shouldActivate: false, insufficientBalance: true};
    }
    const base = live?.expiresAt ?? params.now;
    return {
      shouldActivate: true,
      startedAt: live?.startedAt ?? params.now,
      expiresAt: new Date(base.getTime() + grantMs),
      extendBoostId: live?.boostId,
    };
  }
}

File: lib/features/boost/domain/services/boost_activation_service.dart
import 'package:mevora/features/boost/domain/entities/boost.dart';

class BoostActivationDecision {
  const BoostActivationDecision._({
    required this.shouldActivate,
    this.alreadyActive = false,
    this.insufficientBalance = false,
    this.startedAt,
    this.expiresAt,
    this.extendBoostId,
  });

  factory BoostActivationDecision.activate({
    required DateTime startedAt,
    required DateTime expiresAt,
    String? extendBoostId,
  }) {
    return BoostActivationDecision._(
      shouldActivate: true,
      startedAt: startedAt,
      expiresAt: expiresAt,
      extendBoostId: extendBoostId,
    );
  }

  factory BoostActivationDecision.alreadyActive() {
    return const BoostActivationDecision._(
      shouldActivate: false,
      alreadyActive: true,
    );
  }

  factory BoostActivationDecision.insufficientBalance() {
    return const BoostActivationDecision._(
      shouldActivate: false,
      insufficientBalance: true,
    );
  }

  final bool shouldActivate;
  final bool alreadyActive;
  final bool insufficientBalance;
  final DateTime? startedAt;
  final DateTime? expiresAt;

  /// When set, extend this existing Boost document instead of creating another.
  final String? extendBoostId;
}

/// Server-side policy for granting Boost time.
///
/// Stacking adds [duration] onto remaining time. Duplicate store transactions
/// must be rejected before this service runs.
class BoostActivationService {
  const BoostActivationService({
    this.allowStacking = true,
    this.duration = const Duration(days: 7),
  });

  final bool allowStacking;
  final Duration duration;

  Boost? activeBoost(Iterable<Boost> boosts, DateTime now) {
    for (final boost in boosts) {
      if (boost.isActiveAt(now)) {
        return boost;
      }
    }
    return null;
  }

  /// Treats expiresAt < now as expired even if status is still `active`.
  bool isExpired(Boost boost, DateTime now) => !boost.isActiveAt(now);

  BoostActivationDecision decide({
    required DateTime now,
    Boost? currentActive,
    int balance = 1,
    Duration? duration,
    bool requireBalance = false,
  }) {
    final grant = duration ?? this.duration;
    if (grant <= Duration.zero) {
      return BoostActivationDecision.insufficientBalance();
    }
    final live =
        currentActive != null && currentActive.isActiveAt(now)
            ? currentActive
            : null;
    if (live != null && !allowStacking) {
      return BoostActivationDecision.alreadyActive();
    }
    if (requireBalance && balance < 1) {
      return BoostActivationDecision.insufficientBalance();
    }
    final base = live?.expiresAt ?? now;
    return BoostActivationDecision.activate(
      startedAt: live?.startedAt ?? now,
      expiresAt: base.add(grant),
      extendBoostId: live?.boostId,
    );
  }
}
